aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,404 @@
+//!
+//! let priv_key = KemPrivateKey::generate(&KYBER512_R3)?;
+//!
+//! // Generate private key bytes to possibly save for later decapsulation
+//! let privkey_raw_bytes = priv_key.as_ref();
+//!
+//! let pub_key = priv_key.compute_public_key()?;
+//!
+//! // Get the public key bytes to send to bob through some encoded
+//! // protocol message.
+//! let pub_key_bytes = pub_key.as_ref();
+//!
+//! let mut ciphertext: Vec<u8> = vec![];
+//!
+//! let retrieved_pub_key = KemPublicKey::new(&KYBER512_R3, pub_key_bytes)?;
+//! let bob_result = retrieved_pub_key.encapsulate(|ct, ss| {
+//!     ciphertext.extend_from_slice(ct);
+//!     // In real applications, a KDF would be applied to derive
+//!     // the session keys from the shared secret. We omit that here.
+//!     Ok(())
+//! });
+//!
+//! // Retrieve private key from stored raw bytes
+//! let retrieved_priv_key = KemPrivateKey::new(&KYBER512_R3, privkey_raw_bytes)?;
+//!
+//! let alice_result = retrieved_priv_key.decapsulate(&mut ciphertext, |ss| {
+//!     // In real applications, a KDF would be applied to derive
+//!     // the session keys from the shared secret. We omit that here.
+//!     Ok(())
+//! });
+//!
+//! # Ok::<(), aws_lc_rs::error::Unspecified>(())
+//! ```
+use crate::{
+    error::{KeyRejected, Unspecified},
+    ptr::LcPtr,
+    ptr::Pointer,
+};
+use aws_lc::{
+    EVP_PKEY_CTX_kem_set_params, EVP_PKEY_CTX_new, EVP_PKEY_CTX_new_id, EVP_PKEY_decapsulate,
+    EVP_PKEY_encapsulate, EVP_PKEY_get_raw_private_key, EVP_PKEY_get_raw_public_key,
+    EVP_PKEY_kem_new_raw_public_key, EVP_PKEY_kem_new_raw_secret_key, EVP_PKEY_keygen,
+    EVP_PKEY_keygen_init, EVP_PKEY, EVP_PKEY_KEM, NID_KYBER512_R3,
+};
+use std::os::raw::c_int;
+use std::ptr::null_mut;
+use zeroize::Zeroize;
+
+const KYBER512_R3_SECRET_KEY_LENGTH: usize = 1632;
+const KYBER512_R3_CIPHERTEXT_LENGTH: usize = 768;
+const KYBER512_R3_PUBLIC_KEY_LENGTH: usize = 800;
+const KYBER512_R3_SHARED_SECRET_LENGTH: usize = 32;
+
+#[allow(non_camel_case_types)]
+#[derive(Clone, Debug, PartialEq)]
+
+enum KemAlgorithmID {
+    Kyber512_R3,
+}
+
+/// A KEM algorithm
+#[derive(Debug, PartialEq)]
+pub struct KemAlgorithm {
+    id: KemAlgorithmID,
+    secret_key_size: usize,
+    public_key_size: usize,
+    ciphertext_size: usize,
+    shared_secret_size: usize,
+}
+
+/// NIST Round 3 iteration of the Kyber-512 algorithm
+pub static KYBER512_R3: KemAlgorithm = KemAlgorithm {
+    id: KemAlgorithmID::Kyber512_R3,
+    secret_key_size: KYBER512_R3_SECRET_KEY_LENGTH,
+    public_key_size: KYBER512_R3_PUBLIC_KEY_LENGTH,
+    ciphertext_size: KYBER512_R3_CIPHERTEXT_LENGTH,
+    shared_secret_size: KYBER512_R3_SHARED_SECRET_LENGTH,
+};
+
+impl KemAlgorithmID {
+    #[inline]
+    fn nid(&self) -> i32 {
+        match self {
+            KemAlgorithmID::Kyber512_R3 => NID_KYBER512_R3,
+        }
+    }
+}
+
+impl KemAlgorithm {
+    #[inline]
+    fn secret_key_size(&self) -> usize {
+        self.secret_key_size
```

Comment:
Can't we directly access these values without the get functions?

Reply:
Yeah but sometimes we have to pass the length of a key by pointer to C functions and it can possibly be overwritten. The getter returns a copy of the integer so the static struct's values are never accidentally changed. Also it shows the programmer that these values shouldn't be modified.

Reply:
We could copy the value into another mutable pointer to be passed into the C function. The static struct's values aren't accidentally changed either way.
```
let mut secret_key_size = alg.secret_key_size;
let mut priv_key_bytes = vec![0u8; secret_key_size];
if 1 != EVP_PKEY_get_raw_private_key(
        *kyber_key,
         priv_key_bytes.as_mut_ptr(),
         &mut secret_key_size,
) {
        return Err(Unspecified);
}
```
I think the immutable variable type in the static structure should already show the programmer that the value can't be modified. The way these structure values are immutable and Rust disallows directly passing an immutable reference to the mutable pointer, sort of implies they can't be accidentally changed.

Reply:
I'm fine either way though. I was just curious about why this was needed, and how the rest of the codebase was taking care of this circumstance.

Reply:
Yeah I see what you're saying. Definitely could do it either way. I guess it just depends what it more idiomatic to Rust, which I'm not really sure of

Reply:
You can see that in [cipher.rs](https://github.com/aws/aws-lc-rs/blob/main/aws-lc-rs/src/cipher.rs#L307-L320) they create getters for the relevant struct fields. So seems like a common pattern in the library to me.